File: rust/fory-core/src/row/row.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use std::collections::BTreeMap;

use crate::error::Error;
use crate::types::{Date, Duration, Timestamp};

use super::reader::{ArrayView, MapView};
use super::writer::{ArrayWriter, MapWriter, ValueWriter};

/// Static Row Format behavior for one schema value.
///
/// This trait is public because `ForyRow` implementations are generated in
/// downstream crates. Most applications should derive `ForyRow` instead of
/// implementing it directly.
#[doc(hidden)]
pub trait RowValue {
    /// Zero-copy projection returned when this value is read.
    type View<'a>;

    /// Natural fixed width, or `None` for an offset-addressed value.
    const FIXED_SIZE: Option<usize>;

    /// Writes exactly one value to its container-selected destination.
    fn write(&self, writer: ValueWriter<'_, '_>) -> Result<(), Error>;

    /// Reads exactly one value from its container-resolved bytes.
    fn read<'a>(bytes: &'a [u8]) -> Result<Self::View<'a>, Error>;

    /// Returns true when this value should set its container null bit.
    fn is_null(&self) -> bool {
        false
    }

    /// Produces the projection for a set null bit.
    fn read_null<'a>() -> Result<Self::View<'a>, Error> {
        Err(Error::invalid_data(
            "null row value cannot be read as a non-optional type",
        ))
    }
}

/// A self-contained Standard Row Format root.
///
/// Derived structs, arrays, and maps implement this marker. Scalar, string,
/// binary, and optional values are field/element values rather than row roots.
pub trait Row: RowValue {}

macro_rules! impl_fixed_row_value {
    ($ty:ty, $size:expr) => {
        impl RowValue for $ty {
            type View<'a> = Self;

            const FIXED_SIZE: Option<usize> = Some($size);

            #[inline(always)]
            fn write(&self, writer: ValueWriter<'_, '_>) -> Result<(), Error> {
                writer.write_bytes(&self.to_le_bytes())
            }

            #[inline(always)]
            fn read(bytes: &[u8]) -> Result<Self, Error> {
                Ok(Self::from_le_bytes(read_fixed(bytes)?))
            }
        }
    };
}

impl RowValue for bool {
    type View<'a> = Self;

    const FIXED_SIZE: Option<usize> = Some(1);

    #[inline(always)]
    fn write(&self, writer: ValueWriter<'_, '_>) -> Result<(), Error> {
        writer.write_bytes(&[u8::from(*self)])
    }

    #[inline(always)]
    fn read(bytes: &[u8]) -> Result<Self, Error> {
        match read_fixed::<1>(bytes)?[0] {
            0 => Ok(false),
            1 => Ok(true),
            _ => Err(Error::invalid_data("row boolean must be encoded as 0 or 1")),
        }
    }
}

impl RowValue for i8 {
    type View<'a> = Self;

    const FIXED_SIZE: Option<usize> = Some(1);

    #[inline(always)]
    fn write(&self, writer: ValueWriter<'_, '_>) -> Result<(), Error> {
        writer.write_bytes(&self.to_le_bytes())
    }

    #[inline(always)]
    fn read(bytes: &[u8]) -> Result<Self, Error> {
        Ok(Self::from_le_bytes(read_fixed(bytes)?))
    }
}

impl_fixed_row_value!(i16, 2);
impl_fixed_row_value!(i32, 4);
impl_fixed_row_value!(i64, 8);
impl_fixed_row_value!(f32, 4);
impl_fixed_row_value!(f64, 8);

impl RowValue for String {
    type View<'a> = &'a str;

    const FIXED_SIZE: Option<usize> = None;

    #[inline(always)]
    fn write(&self, writer: ValueWriter<'_, '_>) -> Result<(), Error> {
        writer.write_bytes(self.as_bytes())
    }

    #[inline]
    fn read(bytes: &[u8]) -> Result<&str, Error> {
        std::str::from_utf8(bytes).map_err(|_| Error::invalid_data("invalid UTF-8 in row string"))
    }
}

impl RowValue for &str {
    type View<'a> = &'a str;

    const FIXED_SIZE: Option<usize> = None;

    #[inline(always)]
    fn write(&self, writer: ValueWriter<'_, '_>) -> Result<(), Error> {
        writer.write_bytes(self.as_bytes())
    }

    #[inline]
    fn read(bytes: &[u8]) -> Result<&str, Error> {
        std::str::from_utf8(bytes).map_err(|_| Error::invalid_data("invalid UTF-8 in row string"))
    }
}

impl RowValue for Vec<u8> {
    type View<'a> = &'a [u8];

    const FIXED_SIZE: Option<usize> = None;

    #[inline(always)]
    fn write(&self, writer: ValueWriter<'_, '_>) -> Result<(), Error> {
        writer.write_bytes(self)
    }

    #[inline(always)]
    fn read(bytes: &[u8]) -> Result<&[u8], Error> {
        Ok(bytes)
    }
}

impl RowValue for &[u8] {
    type View<'a> = &'a [u8];

    const FIXED_SIZE: Option<usize> = None;

    #[inline(always)]
    fn write(&self, writer: ValueWriter<'_, '_>) -> Result<(), Error> {
        writer.write_bytes(self)
    }

    #[inline(always)]
    fn read(bytes: &[u8]) -> Result<&[u8], Error> {
        Ok(bytes)
    }
}

impl<T: RowValue> RowValue for Option<T> {
    type View<'a> = Option<T::View<'a>>;

    const FIXED_SIZE: Option<usize> = T::FIXED_SIZE;

    #[inline(always)]
    fn write(&self, writer: ValueWriter<'_, '_>) -> Result<(), Error> {
        match self {
            Some(value) => value.write(writer),
            None => Err(Error::invalid_data(
                "a null row value must be written by its container",
            )),
        }
    }

    #[inline(always)]
    fn read<'a>(bytes: &'a [u8]) -> Result<Self::View<'a>, Error> {
        T::read(bytes).map(Some)
    }

    #[inline(always)]
    fn is_null(&self) -> bool {
        self.is_none()
    }

    #[inline(always)]
    fn read_null<'a>() -> Result<Self::View<'a>, Error> {
        Ok(None)
    }
}

impl RowValue for Date {
    type View<'a> = Self;

    const FIXED_SIZE: Option<usize> = Some(4);

    fn write(&self, writer: ValueWriter<'_, '_>) -> Result<(), Error> {
        let days = i32::try_from(self.epoch_days()).map_err(|_| {
            Error::invalid_data(format!(
                "row date day count {} exceeds date32 range",
                self.epoch_days()
            ))
        })?;
        writer.write_bytes(&days.to_le_bytes())
    }

    fn read(bytes: &[u8]) -> Result<Self, Error> {
        let days = i32::from_le_bytes(read_fixed(bytes)?);
        Ok(Date::from_epoch_days(i64::from(days)))
    }
}

impl RowValue for Timestamp {
    type View<'a> = Self;

    const FIXED_SIZE: Option<usize> = Some(8);

    fn write(&self, writer: ValueWriter<'_, '_>) -> Result<(), Error> {
        writer.write_bytes(&self.to_epoch_micros()?.to_le_bytes())
    }

    fn read(bytes: &[u8]) -> Result<Self, Error> {
        Ok(Timestamp::from_epoch_micros(i64::from_le_bytes(
            read_fixed(bytes)?,
        )))
    }
}

impl RowValue for Duration {
    type View<'a> = Self;

    const FIXED_SIZE: Option<usize> = Some(8);

    fn write(&self, writer: ValueWriter<'_, '_>) -> Result<(), Error> {
        writer.write_bytes(&self.to_micros()?.to_le_bytes())
    }

    fn read(bytes: &[u8]) -> Result<Self, Error> {
        Ok(Duration::from_micros(i64::from_le_bytes(read_fixed(
            bytes,
        )?)))
    }
}

impl<T: RowValue, const N: usize> RowValue for [T; N] {
    type View<'a> = ArrayView<'a, T>;

    const FIXED_SIZE: Option<usize> = None;

    fn write(&self, writer: ValueWriter<'_, '_>) -> Result<(), Error> {
        let mut array_writer = ArrayWriter::<T>::new(N, writer.into_variable()?)?;
        for (index, value) in self.iter().enumerate() {
            array_writer.write(index, value)?;
        }
        Ok(())
    }

    fn read(bytes: &[u8]) -> Result<Self::View<'_>, Error> {
        let view = ArrayView::new(bytes)?;
        if view.len() != N {
            return Err(Error::invalid_data(format!(
                "row fixed array expected {N} elements, found {}",
                view.len()
            )));
        }
        Ok(view)
    }
}

impl<T: RowValue, const N: usize> Row for [T; N] {}

impl<T: RowValue> RowValue for Vec<T> {
    type View<'a> = ArrayView<'a, T>;

    const FIXED_SIZE: Option<usize> = None;

    fn write(&self, writer: ValueWriter<'_, '_>) -> Result<(), Error> {
        let mut array_writer = ArrayWriter::<T>::new(self.len(), writer.into_variable()?)?;
        for (index, value) in self.iter().enumerate() {
            array_writer.write(index, value)?;
        }
        Ok(())
    }

    fn read(bytes: &[u8]) -> Result<Self::View<'_>, Error> {
        ArrayView::new(bytes)
    }
}

impl<T: RowValue> Row for Vec<T> {}

impl<K, V> RowValue for BTreeMap<K, V>
where
    K: RowValue + Ord,
    V: RowValue,
{
    type View<'a> = MapView<'a, K, V>;

    const FIXED_SIZE: Option<usize> = None;

    fn write(&self, writer: ValueWriter<'_, '_>) -> Result<(), Error> {
        let mut map_writer = MapWriter::new(writer.into_variable()?);
        map_writer.write(self)
    }

    fn read(bytes: &[u8]) -> Result<Self::View<'_>, Error> {
        MapView::new(bytes)
    }
}

impl<K, V> Row for BTreeMap<K, V>
where
    K: RowValue + Ord,
    V: RowValue,
{
}

fn read_fixed<const N: usize>(bytes: &[u8]) -> Result<[u8; N], Error> {
    if bytes.len() != N {
        return Err(Error::invalid_data(format!(
            "row fixed-width value expected {N} bytes, found {}",
            bytes.len()
        )));
    }
    let mut value = [0u8; N];
    value.copy_from_slice(bytes);
    Ok(value)
}

File: go/fory/row/row.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package row

import (
	"encoding/binary"
	"fmt"
	"math"
	"time"

	fory "github.com/apache/fory/go/fory"
)

// Row is a zero-copy reader over one row's bytes. Getters read directly
// from the underlying data without deserializing other fields; nested
// values are sub-slice views, never copies (except String).
//
// Row and its nested views require trusted, schema-matched bytes: the
// row format is a trusted in-memory format and these views do not
// validate the row graph. Fixed-width getters return the zero value for
// null fields; use IsNullAt to distinguish. Variable-width getters
// return nil (or "") for null fields. Out-of-range indexes and
// malformed bytes panic.
type Row struct {
	fields      []Field
	data        []byte
	bitmapWidth int
}

func NewRow(schema *Schema, data []byte) *Row {
	return newRow(schema.Fields(), data)
}

func newRow(fields []Field, data []byte) *Row {
	return &Row{fields: fields, data: data, bitmapWidth: bitmapWidthInBytes(len(fields))}
}

func (r *Row) NumFields() int { return len(r.fields) }
func (r *Row) SizeBytes() int { return len(r.data) }

// Data returns the underlying row bytes; callers must not modify them.
func (r *Row) Data() []byte { return r.data }

func (r *Row) IsNullAt(i int) bool {
	if uint(i) >= uint(len(r.fields)) {
		panic(fmt.Sprintf("row: field index %d out of range [0, %d)", i, len(r.fields)))
	}
	return getBit(r.data, i)
}

func (r *Row) slot(i int) int { return r.bitmapWidth + i*8 }

func (r *Row) Bool(i int) bool {
	return !r.IsNullAt(i) && r.data[r.slot(i)] != 0
}

func (r *Row) Int8(i int) int8 {
	if r.IsNullAt(i) {
		return 0
	}
	return int8(r.data[r.slot(i)])
}

func (r *Row) Int16(i int) int16 {
	if r.IsNullAt(i) {
		return 0
	}
	return int16(binary.LittleEndian.Uint16(r.data[r.slot(i):]))
}

func (r *Row) Int32(i int) int32 {
	if r.IsNullAt(i) {
		return 0
	}
	return int32(binary.LittleEndian.Uint32(r.data[r.slot(i):]))
}

func (r *Row) Int64(i int) int64 {
	if r.IsNullAt(i) {
		return 0
	}
	return int64(binary.LittleEndian.Uint64(r.data[r.slot(i):]))
}

func (r *Row) Float32(i int) float32 {
	if r.IsNullAt(i) {
		return 0
	}
	return math.Float32frombits(binary.LittleEndian.Uint32(r.data[r.slot(i):]))
}

func (r *Row) Float64(i int) float64 {
	if r.IsNullAt(i) {
		return 0
	}
	return math.Float64frombits(binary.LittleEndian.Uint64(r.data[r.slot(i):]))
}

func (r *Row) Date(i int) fory.Date {
	if r.IsNullAt(i) {
		return fory.Date{}
	}
	return dateFromDays(int32(binary.LittleEndian.Uint32(r.data[r.slot(i):])))
}

func (r *Row) Timestamp(i int) time.Time {
	if r.IsNullAt(i) {
		return time.Time{}
	}
	return time.UnixMicro(r.Int64(i))
}

func (r *Row) Duration(i int) time.Duration {
	return durationFromMicros(r.Int64(i))
}

// varData returns the value bytes of variable-width field i, or nil if
// the field is null. An empty value returns an empty non-nil slice.
func (r *Row) varData(i int) []byte {
	if r.IsNullAt(i) {
		return nil
	}
	offset, size := decodeOffsetAndSize(binary.LittleEndian.Uint64(r.data[r.slot(i):]))
	return boundedSlice(r.data, offset, size)
}

// Binary returns a zero-copy view of field i's bytes.
func (r *Row) Binary(i int) []byte { return r.varData(i) }

func (r *Row) String(i int) string { return string(r.varData(i)) }

func (r *Row) Struct(i int) *Row {
	data := r.varData(i)
	if data == nil {
		return nil
	}
	structType := r.fields[i].Type.(*StructType)
	return newRow(structType.Fields, data)
}

func (r *Row) Array(i int) *ArrayData {
	data := r.varData(i)
	if data == nil {
		return nil
	}
	listType := r.fields[i].Type.(*ListType)
	return NewArrayData(listType.Elem, data)
}

func (r *Row) Map(i int) *MapData {
	data := r.varData(i)
	if data == nil {
		return nil
	}
	return NewMapData(r.fields[i].Type.(*MapType), data)
}

// ArrayData is a zero-copy reader over one array's bytes, with the same
// null and bounds semantics as Row.
type ArrayData struct {
	elem        Field
	data        []byte
	numElements int
	elemSize    int
	headerBytes int
}

func NewArrayData(elem Field, data []byte) *ArrayData {
	// Narrow the wire count only when it is far below int range, so
	// header and size arithmetic cannot overflow on any platform; an
	// absurd count becomes -1, which validateBounds and every index
	// check reject.
	numElements := -1
	if count := binary.LittleEndian.Uint64(data); count <= uint64(math.MaxInt/8) {
		numElements = int(count)
	}
	elemSize := elem.Type.ByteWidth()
	if elemSize < 0 {
		elemSize = 8
	}
	return &ArrayData{
		elem:        elem,
		data:        data,
		numElements: numElements,
		elemSize:    elemSize,
		headerBytes: 8 + bitmapWidthInBytes(numElements),
	}
}

func (a *ArrayData) NumElements() int { return a.numElements }
func (a *ArrayData) SizeBytes() int   { return len(a.data) }

// validateBounds rejects arrays whose declared element count is not
// covered by the available bytes, so decoders can check before
// allocating from an attacker-declared count.
func (a *ArrayData) validateBounds() error {
	// The coarse count check involves no arithmetic, so it cannot be
	// defeated by overflow: every element occupies at least one data
	// byte. It also bounds numElements low enough that headerBytes
	// (computed in NewArrayData) and the product below are exact.
	if a.numElements < 0 || a.numElements > len(a.data) {
		return fmt.Errorf("row: array declares %d elements but holds only %d bytes", a.numElements, len(a.data))
	}
	need := int64(a.headerBytes) + int64(a.numElements)*int64(a.elemSize)
	if need > int64(len(a.data)) {
		return fmt.Errorf("row: array declares %d elements needing %d bytes but holds only %d", a.numElements, need, len(a.data))
	}
	return nil
}

func (a *ArrayData) IsNullAt(i int) bool {
	if uint(i) >= uint(a.numElements) {
		panic(fmt.Sprintf("row: array index %d out of range [0, %d)", i, a.numElements))
	}
	return getBit(a.data[8:a.headerBytes], i)
}

func (a *ArrayData) slot(i int) int { return a.headerBytes + i*a.elemSize }

func (a *ArrayData) Bool(i int) bool {
	return !a.IsNullAt(i) && a.data[a.slot(i)] != 0
}

func (a *ArrayData) Int8(i int) int8 {
	if a.IsNullAt(i) {
		return 0
	}
	return int8(a.data[a.slot(i)])
}

func (a *ArrayData) Int16(i int) int16 {
	if a.IsNullAt(i) {
		return 0
	}
	return int16(binary.LittleEndian.Uint16(a.data[a.slot(i):]))
}

func (a *ArrayData) Int32(i int) int32 {
	if a.IsNullAt(i) {
		return 0
	}
	return int32(binary.LittleEndian.Uint32(a.data[a.slot(i):]))
}

func (a *ArrayData) Int64(i int) int64 {
	if a.IsNullAt(i) {
		return 0
	}
	return int64(binary.LittleEndian.Uint64(a.data[a.slot(i):]))
}

func (a *ArrayData) Float32(i int) float32 {
	if a.IsNullAt(i) {
		return 0
	}
	return math.Float32frombits(binary.LittleEndian.Uint32(a.data[a.slot(i):]))
}

func (a *ArrayData) Float64(i int) float64 {
	if a.IsNullAt(i) {
		return 0
	}
	return math.Float64frombits(binary.LittleEndian.Uint64(a.data[a.slot(i):]))
}

func (a *ArrayData) Date(i int) fory.Date {
	if a.IsNullAt(i) {
		return fory.Date{}
	}
	return dateFromDays(int32(binary.LittleEndian.Uint32(a.data[a.slot(i):])))
}

func (a *ArrayData) Timestamp(i int) time.Time {
	if a.IsNullAt(i) {
		return time.Time{}
	}
	return time.UnixMicro(a.Int64(i))
}

func (a *ArrayData) Duration(i int) time.Duration {
	return durationFromMicros(a.Int64(i))
}

func (a *ArrayData) varData(i int) []byte {
	if a.IsNullAt(i) {
		return nil
	}
	offset, size := decodeOffsetAndSize(binary.LittleEndian.Uint64(a.data[a.slot(i):]))
	return boundedSlice(a.data, offset, size)
}

// Binary returns a zero-copy view of element i's bytes.
func (a *ArrayData) Binary(i int) []byte { return a.varData(i) }

func (a *ArrayData) String(i int) string { return string(a.varData(i)) }

func (a *ArrayData) Struct(i int) *Row {
	data := a.varData(i)
	if data == nil {
		return nil
	}
	return newRow(a.elem.Type.(*StructType).Fields, data)
}

func (a *ArrayData) Array(i int) *ArrayData {
	data := a.varData(i)
	if data == nil {
		return nil
	}
	return NewArrayData(a.elem.Type.(*ListType).Elem, data)
}

func (a *ArrayData) Map(i int) *MapData {
	data := a.varData(i)
	if data == nil {
		return nil
	}
	return NewMapData(a.elem.Type.(*MapType), data)
}

// MapData is a zero-copy reader over one map's bytes: the keys array
// and values array views share the map's underlying data.
type MapData struct {
	keys   *ArrayData
	values *ArrayData
}

func NewMapData(mapType *MapType, data []byte) *MapData {
	keysSize64 := binary.LittleEndian.Uint64(data)
	if keysSize64 > uint64(len(data)) {
		panic(fmt.Sprintf("row: map keys array of %d bytes exceeds the %d-byte map region", keysSize64, len(data)))
	}
	keysSize := int(keysSize64)
	keysData := boundedSlice(data, 8, keysSize)
	valuesData := boundedSlice(data, 8+keysSize, len(data)-8-keysSize)
	return &MapData{
		keys:   NewArrayData(mapType.Key, keysData),
		values: NewArrayData(mapType.Value, valuesData),
	}
}

// boundedSlice checks declared bounds against the data LENGTH
// and caps the view so nested
// readers cannot reach outside it either. Out-of-bounds panics are
// converted to errors by the encoder's decode entry points.
func boundedSlice(data []byte, offset, size int) []byte {
	end := offset + size
	if offset < 0 || size < 0 || end > len(data) {
		panic(fmt.Sprintf("row: value bytes [%d:%d] exceed the enclosing %d-byte region", offset, end, len(data)))
	}
	return data[offset:end:end]
}

func (m *MapData) NumElements() int   { return m.keys.numElements }
func (m *MapData) Keys() *ArrayData   { return m.keys }
func (m *MapData) Values() *ArrayData { return m.values }

func decodeOffsetAndSize(slotValue uint64) (offset, size int) {
	return int(slotValue >> 32), int(uint32(slotValue))
}

// dateFromDays converts int32 epoch days to a Date. The int32 range
// always yields an in-range year, so the conversion cannot fail.
func dateFromDays(days int32) fory.Date {
	d, _ := fory.DateFromEpochDay(int64(days))
	return d
}

// durationFromMicros converts wire microseconds to a time.Duration,
// which counts nanoseconds; values the multiplication would wrap are
// rejected instead of silently changing sign.
func durationFromMicros(micros int64) time.Duration {
	if micros > math.MaxInt64/1000 || micros < math.MinInt64/1000 {
		panic(fmt.Sprintf("row: duration of %d microseconds overflows time.Duration", micros))
	}
	return time.Duration(micros) * time.Microsecond
}
